forge-game/src/main/java/forge/game/ability/effects/DiscoverEffect.java

```diff
@@ -0,0 +1,179 @@
+import forge.game.trigger.TriggerType;
+import forge.game.zone.PlayerZone;
+import forge.game.zone.Zone;
+import forge.game.zone.ZoneType;
+import forge.util.CardTranslation;
+import forge.util.Localizer;
+
+import java.util.List;
+import java.util.Map;
+
+public class DiscoverEffect extends SpellAbilityEffect {
+
+    @Override
+    protected String getStackDescription(SpellAbility sa) {
+        final StringBuilder sb = new StringBuilder();
+        sb.append(sa.getActivatingPlayer()).append(" discovers ");
+        sb.append(sa.getParamOrDefault("Num", "1")).append(".");
+
+        return sb.toString();
+    }
+
+    @Override
+    public void resolve(SpellAbility sa) {
+        final Card host = sa.getHostCard();
+        final Game game = host.getGame();
+        final Player p = sa.getActivatingPlayer();
+
+        // Exile cards from the top of your library until you exile a nonland card with <N> mana value or less.
+        final String val = sa.getParamOrDefault("Num", "1");
+        final int num = AbilityUtils.calculateAmount(host, val, sa);
+
+        if (p == null || !p.isInGame()) return;
+
+        Card found = null;
+        CardCollection exiled = new CardCollection();
+        CardCollection rest = new CardCollection();
+
+        final PlayerZone library = p.getZone(ZoneType.Library);
+
+        for (int i = 0; i < library.size(); i++) {
+            final Card c = library.get(i);
+            exiled.add(c);
+            if (c.isValid("Card.nonLand+cmcLE" + val, sa.getActivatingPlayer(), host, sa)) {
+                found = c;
+                if (sa.hasParam("RememberDiscovered"))
+                    host.addRemembered(c);
+                break;
+            } else {
+                rest.add(c);
+            }
+        }
+
+        if (exiled.size() > 0) {
+            game.getAction().reveal(exiled, p, false);
+        }
+
+        changeZone(exiled, ZoneType.Exile, game, sa);
+
+        // Cast it without paying its mana cost or put it into your hand.
+        if (found != null) {
+            String prompt = Localizer.getInstance().getMessage("lblDiscoverChoice",
+                            CardTranslation.getTranslatedName(found.getName()));
+            final Zone origin = found.getZone();
+            final boolean play = p.getController().chooseBinary(sa, prompt,
+                    PlayerController.BinaryChoiceType.CastOrHand);
+            boolean cancel = false;
+
+            if (play) {
+                // get basic spells (no flashback, etc.)
+                List<SpellAbility> sas = AbilityUtils.getBasicSpellsFromPlayEffect(found, p);
+
+                // filter out land abilities due to MDFC or similar
+                Iterables.removeIf(sas, Predicates.instanceOf(LandAbility.class));
+                // the spell must also have a mana value equal to or less than the discover number
+                sas.removeIf(sp -> !sp.isValid("Spell.cmcLE" + val, p , host, sa));
```

Comment:
same with there, the SpellCMC should be filtered out better than using isValid?

have you tested it with Adventure spells? if the check really gives the right cmc?